Add blog article: Kubernetes v1.26: Advancements in Kubernetes Traffic Engineering (#37648)
kubernetes/website: content/en/blog/_posts/2022-11-28-advancements-in-traffic-engineering/index.md

---
layout: blog
title: "Kubernetes v1.26: Advancements in Kubernetes Traffic Engineering"

> date: 2022-11-28

Reviewer: ```suggestion
date: 2022-12-30
```
Please also change the path `content/en/blog/_posts/2022-11-28-advancements-in-traffic-engineering/index.md` to match.

Author: Updated publication date, thanks

---
layout: blog
title: "Kubernetes v1.26: Advancements in Kubernetes Traffic Engineering"
date: 2022-11-28
slug: advancements-in-kubernetes-traffic-engineering
---

**Authors:** Andrew Sy Kim (Google)

> Kubernetes v1.26 includes significant advancements in traffic engineering with the graduation of the
> `ServiceInternalTrafficPolicy` and `EndpointSliceTerminatingCondition` features to GA and the
> `ProxyTerminatingEndpoints` feature to Beta. The combination of these enhancements aim to address
> short-comings in traffic engineering today and unlock new capabilities for the future.

Reviewer: ```suggestion
Kubernetes v1.26 includes significant advancements in network traffic engineering with the graduation of
two features (Service internal traffic policy support, and EndpointSlice terminating conditions) GA,
and a third feature (Proxy terminating endpoints) `ProxyTerminatingEndpoints` feature to beta.
The combination of these enhancements aims to address short-comings in traffic engineering that
people face today, and unlock new capabilities for the future.
```

No need to name the feature gates in the introduction; better to let people know that detail later,
if needed.

## Traffic Loss from Load Balancers During Rolling Updates

> Prior to Kubernetes v1.26, [clusters can experience loss of traffic](https://github.com/kubernetes/kubernetes/issues/85643)

Reviewer: ```suggestion
Prior to Kubernetes v1.26, clusters could experience [loss of traffic](https://github.com/kubernetes/kubernetes/issues/85643)
```

from Service load balancers during rolling updates when setting the `externalTrafficPolicy` field to `Local`.
There are a lot of moving parts at play here so a quick overview of how Kubernetes manages load balancers might help!

> In Kubernetes, you can create a Service with Type=LoadBalancer to expose an application externally with a load balancer.

Reviewer: ```suggestion
In Kubernetes, you can create a Service with `type: LoadBalancer` to expose an application externally with a load balancer.
```
YAML is the norm; JSON is a fallback. `Type=LoadBalancer` isn't part of the API either way (it's more like Golang).

The load balancer implementation varies between clusters and platforms, but the Service provides a generic abstraction
representing the load balancer that is consistent across all Kubernetes installations.

```yaml
apiVersion: v1
kind: Service
metadata:
  name: my-service
spec:
  selector:
    app.kubernetes.io/name: my-app
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
  type: LoadBalancer
```

Under the hood, Kubernetes allocates a NodePort for the Service, which is then used by kube-proxy to provide a

> network data path from the NodePort to the Pod. A controller running in the Kubernetes Control Plane will then

Reviewer: ```suggestion
network data path from the NodePort to the Pod. A controller will then
```
The boundary is blurred here; you can set `service.kubernetes.io/service-proxy-name` on a Service and use your own load balancer. We can assume that the thing watching for Services is a controller, though.

add all available Nodes in the cluster to the load balancer’s backend pool, using the designated NodePort for
the Service as the backend target port.

> {{< figure src="traffic-engineering-service-load-balancer.png" alt="Figure 1: Overview of Service Load Balancers" >}}

Reviewer: (nit) You can caption the image rather than setting the figure name only in the `alt` attribute.

Oftentimes it is beneficial to set `externalTrafficPolicy: Local` for Services, to avoid extra hops between
Nodes that are not running healthy Pods backing that Service. When using `externalTrafficPolicy: Local`,
an additional NodePort is allocated for health checking purposes, such that Nodes that do not contain healthy
Pods are not included in the backend pool for a load balancer.

{{< figure src="traffic-engineering-lb-healthy.png" alt="" >}}

One such scenario where traffic can be lost is when a Node loses the last remaining Pod for a Service,
but the external load balancer has not probed the health check NodePort yet. The likelihood of this situation
is largely dependent on the health checking interval configured on the load balancer. The larger the interval,
the more likely this will happen, since the load balancer will take longer to realize that a node is no longer
healthy and unable to accept traffic. This also occurred when Pods started terminating during rolling updates,
specifically when a Node went from N to 0 healthy pods because Kubernetes does not consider terminating Pods

> as “Ready”, regardless of their readiness probes.

Reviewer: ```suggestion
as `Ready`, regardless of their readiness probes.
```

{{< figure src="traffic-engineering-lb-unhealthy.png" alt="" >}}

> Starting in Kubernetes v1.26, kube-proxy enables the `ProxyTerminatingEndpoints` feature by default, which
> adds automatic failover and routing to terminating endpoints in scenarios where the traffic would otherwise
> be dropped. More specifically, when there is a rolling update and a Node only contains terminating replicas,
> kube-proxy will route traffic to the terminating replicas as long as their readiness probes are passing.
> By doing so, kube-proxy provides the external load balancer a window of time to gracefully steer traffic
> away from the Node after its next health check probe.

Reviewer: Is this right:
```suggestion
Starting in Kubernetes v1.26, kube-proxy defaults to using _proxy terminating endpoints_ behavior. The
kube-proxy spots when traffic arrives that is destined to a terminating endpoint, and avoids some scenarios where the traffic would otherwise be dropped.

More specifically, when there is a rolling update and a Node only contains terminating replicas,
kube-proxy routes traffic to the terminating replicas only as long as their readiness probes are passing.
By doing so, kube-proxy provides the external load balancer a window of time to gracefully steer traffic
away from the Node after its next health check probe.
If there are no local healthy replicas, kube-proxy looks for another node with healthy endpoints and
tries to send the traffic there.
```
?

I'm worried that readers won't work out how the new behavior is different.
